Add maintenace docs (#232)
elastisys/compliantkubernetes: docs/compliantkubernetes/operator-manual/maintenance.md

# Maintaining and Upgrading your Compliant Kubernetes environment.

> In order to keep your Compliant Kubernetes environment running smoothly, and to assure that you are up to date with the latest patches you need to perform regular maintenance on ut.

Reviewer: ```suggestion
In order to keep your Compliant Kubernetes environment running smoothly, and to assure that you are up to date with the latest security patches you need to perform regular maintenance on it.
```

This guide assumes that:

> - Your Compliant Kubernetes environment is running normally, if not, please see the [Troubleshooting guide](https://elastisys.io/compliantkubernetes/operator-manual/troubleshooting/).

Reviewer: ```suggestion
- Your Compliant Kubernetes environment is running normally, if not, please see the [Troubleshooting guide](/compliantkubernetes/operator-manual/troubleshooting/).
```

> - Your Compliant Kubernetes environment is properly [sized](https://elastisys.io/compliantkubernetes/operator-manual/cluster-sizing/)

Reviewer: ```suggestion
- Your Compliant Kubernetes environment is properly sized; see [cluster sizing](/compliantkubernetes/operator-manual/cluster-sizing/) and [capacity management](/compliantkubernetes/operator-manual/capacity-management/)
```

> - You have performed the actions in the [Go-live Checklist](https://elastisys.io/compliantkubernetes/user-guide/go-live/) as failure to do so might cause downtime during maintenance.

Reviewer: ```suggestion
- You have performed the actions in the [Go-live Checklist](/compliantkubernetes/user-guide/go-live/) as failure to do so might cause downtime during maintenance.
```

## Compliance needs

> Many regulations require you to secure your information system against unathorized access, data loss, and breaches.

Reviewer: ```suggestion
Many regulations require you to secure your information system against unauthorized access, data loss, and breaches.
```

3. If there is no relevant migration document, first do a dry-run.

    ```bash
    ./bin/ck8s dry-run sc
    ./bin/ck8s dry-run wc

4. If dry-run reports no errors, proceed with the upgrade.

    ```bash
    ./bin/ck8s apply sc
    ./bin/ck8s apply wc
    ```

5. Verify that everything is running after the upgrade.
At the minimum, at least run the tests in compliantkubernetes-apps.

    ```bash
    ./bin/ck8s test sc
    ./bin/ck8s test wc
    ```

6. Go back to step 1 and repeat one new release of compliantkubernetes-apps at a time until you are at the latest release.

### Upgrading Kubespray/Kubernetes

All clusters should stay up to date with the latest Kubespray version used in [compliantkubernetes-kubespray](https://github.com/elastisys/compliantkubernetes-kubespray).

1. Note what version of Kubespray that is currently used in the cluster and the Kubespray version we want to upgrade to.
Then check the release notes for each version in between to see if there are anything that might cause any problems, if so then consult the rest of the operations team before proceeding.

> Also check if the newer Kubespray version would upgrade Kubernetes to a new minor version, if so then the customer should get a notice of x weeks before proceeding to let them check for any deprecated APIs that they might be using.

Reviewer: We tend to use "user" instead of "customer" in this documentation, but it really means the same thing. Please decide if you prefer "community" terminology or "business" terminology. :smile: (I'm tempted to add "customer" to `vocabulary.md`).

Patches that are released to an older minor version can be skipped, e.g. new patches to 2.14 after 2.15 has been released.
Read more about Kubespray upgrades in their [documentation](https://kubespray.io/#/docs/upgrades).

1. Checkout the next Kubespray version by checking out the last compliantkubernetes-kubespray commit (the commit is `next-version` in the snippet below) that used that version and updating the submodule.

    ```bash
    # you should be in the root folder of compliantkubernetes-kubespray
    git checkout <next-version>
    git submodule update
    ```

2. Upgrade compliantkubernetes-kubespray by following the relevant [documentation](https://github.com/elastisys/compliantkubernetes-kubespray/tree/main/migration) (e.g. [for upgrade to v2.17.x-ck8s1](https://github.com/elastisys/compliantkubernetes-kubespray/blob/v2.17.1-ck8s1/migration/v2.16.0-ck8s1-v2.17.x-ck8s1/upgrade-cluster.md)).

## After doing any upgrades or maintenance

It is a good idea to perform a **log review** after each maintenance.
The purpose is two-fold:

1. To catch potential issues caused by the maintenance.
2. To piggy-back log review on top of a known-good process.

> For instructions on how to do this, please follow the [log review guide](https://compliantkubernetes.io/ciso-guide/log-review/).

Reviewer: Adjusting URL to new location:

```suggestion
For instructions on how to do this, please follow the [log review guide](/compliantkubernetes.io/ciso-guide/log-review/).
```
